internal/resolver/dns/dns_resolver_test.go

```diff
@@ -57,13 +58,25 @@ type testClientConn struct {
 	state               resolver.State
 	updateStateCalls    int
 	errChan             chan error
+
+	backoffMutex              sync.Mutex
+	causeDNSResolverToBackoff bool
```

Comment:
Better to name this after what it does than what side-effects it has.

How about `updateStateErr error` (and return this error always - if `nil`, that's success).

That you need a mutex here is a little concerning.  I'd probably use a channel of errors instead, but that might require more significant refactoring elsewhere.  So, consider it, but this is fine if that's hard.

Note that we don't have to cram all functionality into _one_ `testClientConn`.  We can make other fake `ClientConn`s, if it helps for clarity or ease of use.

Reply:
Changed whole "causeDNSResolverToBackoff" logic to an error that is part of testClientConn state.